File: packages/offscreen-webview-sdk/native/ios/OffscreenWebViewModule.swift
import Foundation
import WebKit
import UIKit

@objc(OffscreenWebViewModule)
class OffscreenWebViewModule: NSObject {
  
  private var webViews: [String: WKWebView] = [:]
  private var eventEmitter: RCTEventEmitter?
  
  override init() {
    super.init()
  }
  
  @objc
  func createInstance(_ instanceId: String, config: [String: Any], resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      do {
        let webView = try self.createWebView(instanceId: instanceId, config: config)
        self.webViews[instanceId] = webView
        resolver(nil)
      } catch {
        rejecter("CREATE_ERROR", "Failed to create WebView instance", error)
      }
    }
  }
  
  @objc
  func loadUrl(_ instanceId: String, url: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      guard let url = URL(string: url) else {
        rejecter("INVALID_URL", "Invalid URL", nil)
        return
      }
      
      let request = URLRequest(url: url)
      webView.load(request)
      resolver(nil)
    }
  }
  
  @objc
  func loadHTML(_ instanceId: String, html: String, baseURL: String?, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      let baseURLValue = baseURL != nil ? URL(string: baseURL!) : nil
      webView.loadHTMLString(html, baseURL: baseURLValue)
      resolver(nil)
    }
  }
  
  @objc
  func executeJavaScript(_ instanceId: String, script: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      webView.evaluateJavaScript(script) { result, error in
        if let error = error {
          rejecter("JS_ERROR", "JavaScript execution failed", error)
        } else {
          resolver(result)
        }
      }
    }
  }
  
  @objc
  func captureScreenshot(_ instanceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      let renderer = UIGraphicsImageRenderer(bounds: webView.bounds)
      let image = renderer.image { context in
        webView.layer.render(in: context.cgContext)
      }
      
      guard let imageData = image.pngData() else {
        rejecter("SCREENSHOT_ERROR", "Failed to create screenshot", nil)
        return
      }
      
      let base64String = imageData.base64EncodedString()
      resolver(base64String)
    }
  }
  
  @objc
  func getPageContent(_ instanceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      webView.evaluateJavaScript("document.documentElement.outerHTML") { result, error in
        if let error = error {
          rejecter("CONTENT_ERROR", "Failed to get page content", error)
        } else {
          resolver(result as? String ?? "")
        }
      }
    }
  }
  
  @objc
  func getPageTitle(_ instanceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      webView.evaluateJavaScript("document.title") { result, error in
        if let error = error {
          rejecter("TITLE_ERROR", "Failed to get page title", error)
        } else {
          resolver(result as? String ?? "")
        }
      }
    }
  }
  
  @objc
  func getCurrentUrl(_ instanceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      resolver(webView.url?.absoluteString ?? "")
    }
  }
  
  @objc
  func goBack(_ instanceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      let canGoBack = webView.canGoBack
      if canGoBack {
        webView.goBack()
      }
      resolver(canGoBack)
    }
  }
  
  @objc
  func goForward(_ instanceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      let canGoForward = webView.canGoForward
      if canGoForward {
        webView.goForward()
      }
      resolver(canGoForward)
    }
  }
  
  @objc
  func reload(_ instanceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      webView.reload()
      resolver(nil)
    }
  }
  
  @objc
  func stopLoading(_ instanceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let webView = webViews[instanceId] else {
      rejecter("NOT_FOUND", "WebView instance not found", nil)
      return
    }
    
    DispatchQueue.main.async {
      webView.stopLoading()
      resolver(nil)
    }
  }
  
  @objc
  func destroy(_ instanceId: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      if let webView = self.webViews[instanceId] {
        webView.stopLoading()
        webView.loadHTMLString("", baseURL: nil)
        self.webViews.removeValue(forKey: instanceId)
      }
      resolver(nil)
    }
  }
  
  // MARK: - Private Methods
  
  private func createWebView(instanceId: String, config: [String: Any]) throws -> WKWebView {
    let configuration = WKWebViewConfiguration()
    
    // 配置 JavaScript 支持
    if let javaScriptEnabled = config["javaScriptEnabled"] as? Bool {
      configuration.preferences.javaScriptEnabled = javaScriptEnabled
    }
    
    // 配置用户代理
    if let userAgent = config["userAgent"] as? String {
      configuration.applicationNameForUserAgent = userAgent
    }
    
    // 配置允许文件访问
    if let allowFileAccess = config["allowFileAccess"] as? Bool {
      configuration.preferences.javaScriptCanOpenWindowsAutomatically = allowFileAccess
    }
    
    // 创建 WebView
    let webView = WKWebView(frame: .zero, configuration: configuration)
    webView.navigationDelegate = self
    webView.uiDelegate = self
    
    // 设置尺寸
    if let width = config["width"] as? CGFloat, let height = config["height"] as? CGFloat {
      webView.frame = CGRect(x: 0, y: 0, width: width, height: height)
    }
    
    // 注入初始 JavaScript
    if let injectedJavaScript = config["injectedJavaScript"] as? String {
      webView.evaluateJavaScript(injectedJavaScript)
    }
    
    return webView
  }
}

// MARK: - WKNavigationDelegate
extension OffscreenWebViewModule: WKNavigationDelegate {
  func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
    // 发送加载开始事件
    sendEvent(webView: webView, eventName: "onLoadStart", data: ["url": webView.url?.absoluteString ?? ""])
  }
  
  func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
    // 发送加载完成事件
    sendEvent(webView: webView, eventName: "onLoadEnd", data: ["url": webView.url?.absoluteString ?? ""])
  }
  
  func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
    // 发送错误事件
    sendEvent(webView: webView, eventName: "onError", data: [
      "url": webView.url?.absoluteString ?? "",
      "error": error.localizedDescription
    ])
  }
  
  func webView(_ webView: WKWebView, didReceiveServerRedirectForProvisionalNavigation navigation: WKNavigation!) {
    // 发送标题变化事件
    webView.evaluateJavaScript("document.title") { result, error in
      if let title = result as? String {
        self.sendEvent(webView: webView, eventName: "onTitleChange", data: ["title": title])
      }
    }
  }
}

// MARK: - WKUIDelegate
extension OffscreenWebViewModule: WKUIDelegate {
  // 处理 JavaScript 弹窗等
}

// MARK: - Event Handling
extension OffscreenWebViewModule {
  private func sendEvent(webView: WKWebView, eventName: String, data: [String: Any]) {
    // 找到对应的 instanceId
    let instanceId = webViews.first { $0.value == webView }?.key ?? ""
    let eventData: [String: Any] = [
      "instanceId": instanceId,
      "data": data
    ]
    
    // 这里需要实现事件发送逻辑
    // 可以通过 RCTEventEmitter 发送事件到 React Native
  }
}

// MARK: - React Native Module Export
@objc
extension OffscreenWebViewModule {
  static func requiresMainQueueSetup() -> Bool {
    return true
  }
} 

File: packages/offscreen-webview-sdk/native/android/OffscreenWebViewModule.java
package com.native.offscreenwebview;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.ValueCallback;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceError;
import android.util.Base64;
import android.util.Log;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.io.ByteArrayOutputStream;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class OffscreenWebViewModule extends ReactContextBaseJavaModule {
    private static final String TAG = "OffscreenWebViewModule";
    private final ReactApplicationContext reactContext;
    private final Map<String, WebView> webViews = new ConcurrentHashMap<>();
    private final Map<String, Promise> pendingPromises = new ConcurrentHashMap<>();

    public OffscreenWebViewModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "OffscreenWebViewModule";
    }

    @ReactMethod
    public void createInstance(String instanceId, ReadableMap config, Promise promise) {
        try {
            WebView webView = createWebView(instanceId, config);
            webViews.put(instanceId, webView);
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("CREATE_ERROR", "Failed to create WebView instance", e);
        }
    }

    @ReactMethod
    public void loadUrl(String instanceId, String url, Promise promise) {
        WebView webView = webViews.get(instanceId);
        if (webView == null) {
            promise.reject("NOT_FOUND", "WebView instance not found");
            return;
        }

        try {
            webView.loadUrl(url);
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("LOAD_ERROR", "Failed to load URL", e);
        }
    }

    @ReactMethod
    public void loadHTML(String instanceId, String html, String baseURL, Promise promise) {
        WebView webView = webViews.get(instanceId);
        if (webView == null) {
            promise.reject("NOT_FOUND", "WebView instance not found");
            return;
        }

        try {
            webView.loadDataWithBaseURL(baseURL, html, "text/html", "UTF-8", null);
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("LOAD_ERROR", "Failed to load HTML", e);
        }
    }

    @ReactMethod
    public void executeJavaScript(String instanceId, String script, Promise promise) {
        WebView webView = webViews.get(instanceId);
        if (webView == null) {
            promise.reject("NOT_FOUND", "WebView instance not found");
            return;
        }

        try {
            webView.evaluateJavascript(script, new ValueCallback<String>() {
                @Override
                public void onReceiveValue(String value) {
                    promise.resolve(value);
                }
            });
        } catch (Exception e) {
            promise.reject("JS_ERROR", "JavaScript execution failed", e);
        }
    }

    @ReactMethod
    public void captureScreenshot(String instanceId, Promise promise) {
        WebView webView = webViews.get(instanceId);
        if (webView == null) {
            promise.reject("NOT_FOUND", "WebView instance not found");
            return;
        }

        try {
            // 创建位图
            Bitmap bitmap = Bitmap.createBitmap(webView.getWidth(), webView.getHeight(), Bitmap.Config.ARGB_8888);
            Canvas canvas = new Canvas(bitmap);
            webView.draw(canvas);

            // 转换为 Base64
            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream);
            byte[] byteArray = byteArrayOutputStream.toByteArray();
            String base64 = Base64.encodeToString(byteArray, Base64.DEFAULT);

            promise.resolve(base64);
        } catch (Exception e) {
            promise.reject("SCREENSHOT_ERROR", "Failed to capture screenshot", e);
        }
    }

    @ReactMethod
    public void getPageContent(String instanceId, Promise promise) {
        executeJavaScript(instanceId, "document.documentElement.outerHTML", promise);
    }

    @ReactMethod
    public void getPageTitle(String instanceId, Promise promise) {
        executeJavaScript(instanceId, "document.title", promise);
    }

    @ReactMethod
    public void getCurrentUrl(String instanceId, Promise promise) {
        WebView webView = webViews.get(instanceId);
        if (webView == null) {
            promise.reject("NOT_FOUND", "WebView instance not found");
            return;
        }

        try {
            String url = webView.getUrl();
            promise.resolve(url != null ? url : "");
        } catch (Exception e) {
            promise.reject("URL_ERROR", "Failed to get current URL", e);
        }
    }

    @ReactMethod
    public void goBack(String instanceId, Promise promise) {
        WebView webView = webViews.get(instanceId);
        if (webView == null) {
            promise.reject("NOT_FOUND", "WebView instance not found");
            return;
        }

        try {
            boolean canGoBack = webView.canGoBack();
            if (canGoBack) {
                webView.goBack();
            }
            promise.resolve(canGoBack);
        } catch (Exception e) {
            promise.reject("NAVIGATION_ERROR", "Failed to go back", e);
        }
    }

    @ReactMethod
    public void goForward(String instanceId, Promise promise) {
        WebView webView = webViews.get(instanceId);
        if (webView == null) {
            promise.reject("NOT_FOUND", "WebView instance not found");
            return;
        }

        try {
            boolean canGoForward = webView.canGoForward();
            if (canGoForward) {
                webView.goForward();
            }
            promise.resolve(canGoForward);
        } catch (Exception e) {
            promise.reject("NAVIGATION_ERROR", "Failed to go forward", e);
        }
    }

    @ReactMethod
    public void reload(String instanceId, Promise promise) {
        WebView webView = webViews.get(instanceId);
        if (webView == null) {
            promise.reject("NOT_FOUND", "WebView instance not found");
            return;
        }

        try {
            webView.reload();
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("RELOAD_ERROR", "Failed to reload", e);
        }
    }

    @ReactMethod
    public void stopLoading(String instanceId, Promise promise) {
        WebView webView = webViews.get(instanceId);
        if (webView == null) {
            promise.reject("NOT_FOUND", "WebView instance not found");
            return;
        }

        try {
            webView.stopLoading();
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("STOP_ERROR", "Failed to stop loading", e);
        }
    }

    @ReactMethod
    public void destroy(String instanceId, Promise promise) {
        try {
            WebView webView = webViews.remove(instanceId);
            if (webView != null) {
                webView.stopLoading();
                webView.loadUrl("about:blank");
                webView.destroy();
            }
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("DESTROY_ERROR", "Failed to destroy WebView", e);
        }
    }

    private WebView createWebView(String instanceId, ReadableMap config) {
        WebView webView = new WebView(reactContext);
        
        // 配置 WebView 设置
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setAllowFileAccess(true);
        settings.setAllowContentAccess(true);
        settings.setLoadsImagesAutomatically(true);
        settings.setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
        
        // 设置用户代理
        if (config.hasKey("userAgent")) {
            settings.setUserAgentString(config.getString("userAgent"));
        }
        
        // 设置尺寸
        if (config.hasKey("width") && config.hasKey("height")) {
            int width = config.getInt("width");
            int height = config.getInt("height");
            webView.setLayoutParams(new android.view.ViewGroup.LayoutParams(width, height));
        }

        // 设置 WebViewClient
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                sendEvent(instanceId, "onLoadStart", createEventData("url", url));
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                sendEvent(instanceId, "onLoadEnd", createEventData("url", url));
            }

            @Override
            public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
                WritableMap errorData = Arguments.createMap();
                errorData.putString("url", request.getUrl().toString());
                errorData.putString("error", error.getDescription().toString());
                sendEvent(instanceId, "onError", errorData);
            }
        });

        // 设置 WebChromeClient
        webView.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onProgressChanged(WebView view, int newProgress) {
                WritableMap progressData = Arguments.createMap();
                progressData.putInt("progress", newProgress);
                sendEvent(instanceId, "onProgress", progressData);
            }

            @Override
            public void onReceivedTitle(WebView view, String title) {
                sendEvent(instanceId, "onTitleChange", createEventData("title", title));
            }
        });

        // 注入初始 JavaScript
        if (config.hasKey("injectedJavaScript")) {
            String injectedScript = config.getString("injectedJavaScript");
            webView.evaluateJavascript(injectedScript, null);
        }

        return webView;
    }

    private WritableMap createEventData(String key, String value) {
        WritableMap data = Arguments.createMap();
        data.putString(key, value);
        return data;
    }

    private void sendEvent(String instanceId, String eventName, WritableMap data) {
        WritableMap eventData = Arguments.createMap();
        eventData.putString("instanceId", instanceId);
        eventData.putMap("data", data);

        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
            .emit(instanceId + ":" + eventName, eventData);
    }
} 